test/jdk/java/awt/Toolkit/ToolkitListenerTest/ToolkitListenerTest.java

```diff
@@ -0,0 +1,117 @@
+ */
+
+/*
+  @test
+  @bug 4460376
+  @summary we should create Component-, Container- and HierarchyEvents if
+  appropriate AWTEventListener added on Toolkit
+  @key headful
+*/
+
+import java.awt.AWTEvent;
+import java.awt.Button;
+import java.awt.EventQueue;
+import java.awt.Frame;
+import java.awt.Toolkit;
+import java.awt.event.AWTEventListener;
+import java.awt.event.ComponentEvent;
+import java.awt.event.ContainerEvent;
+import java.awt.event.HierarchyEvent;
+import java.lang.reflect.InvocationTargetException;
+
+public class ToolkitListenerTest implements AWTEventListener
+{
+    public static Frame frame;
+    static boolean containerEventReceived = false;
+    static boolean componentEventReceived = false;
+    static boolean hierarchyEventReceived = false;
+    static boolean hierarchyBoundsEventReceived = false;
+
+    public static void main(String[] args) throws Exception {
+        ToolkitListenerTest test = new ToolkitListenerTest();
+        test.start();
+    }
+    public void start() throws Exception {
+        Toolkit.getDefaultToolkit().
+            addAWTEventListener(this,
+                AWTEvent.COMPONENT_EVENT_MASK |
+                    AWTEvent.CONTAINER_EVENT_MASK |
+                    AWTEvent.HIERARCHY_EVENT_MASK |
+                    AWTEvent.HIERARCHY_BOUNDS_EVENT_MASK);
+        EventQueue.invokeAndWait(() -> {
+            frame = new Frame("ToolkitListenerTest");
+            frame.setSize(200,200);
+            frame.add(new Button());
+            frame.setBounds(100, 100, 100, 100);
+        });
+        try {
+            Toolkit.getDefaultToolkit().getSystemEventQueue().
+                invokeAndWait(new Runnable() {
+                    public void run() {}
+                });
+
+            if (!componentEventReceived) {
+                throw new RuntimeException("Test Failed: ComponentEvent was not dispatched");
+            }
+            if (!containerEventReceived) {
+                throw new RuntimeException("Test Failed: ContainerEvent was not dispatched");
+            }
+            if (!hierarchyEventReceived) {
+                throw new RuntimeException("Test Failed: HierarchyEvent(HIERARCHY_CHANGED) was not dispatched");
+            }
+            if (!hierarchyBoundsEventReceived) {
+                throw new RuntimeException("Test Failed: HierarchyEvent(ANCESTOR_MOVED or ANCESTOR_RESIZED) was not dispatched");
+            }
+        } catch (InterruptedException ie) {
+            throw new RuntimeException("Test Failed: InterruptedException accured.");
+        } catch (InvocationTargetException ite) {
+            throw new RuntimeException("Test Failed: InvocationTargetException accured.");
+        } finally {
+            EventQueue.invokeAndWait(() -> {
+                if (frame != null) {
+                    frame.dispose();
+                }
+            });
+        }
+    }
+
+    public void eventDispatched(AWTEvent e) {
+        System.err.println(e);
```

Comment:
On what thread this will be executed?

Reply:
Surely it will be the EDT ? Why do you ask ?

Reply:
it sets a couple of flags that will be used on the main thread, see "start()" above

Reply:
I added logs and saw that before we check these flags we have synchronous EDT tasks, that is the reason it was not failing. But yes we need to check these flags in EDT only. @mrserb Now i am checking these flags in EDT.